### PR TITLE
Alarmrefactor

### PR DESCRIPTION
**Pull request by PatrickGe**
complete refactor of the alarm
#326 
Please also have a look at: https://github.com/AmyAssist/Amy-Web/pull/67
(might be still in progress but will finish soon)

Pull request status:
- [x] Tests added/updated
- [x] 71 % Coverage because it will be overwritten in the next Sprint
- [x] Code conventions
- [x] Ready for review

### CODE REVIEW

File: plugins/alarmclock/src/test/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmBeepServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+/**
+ * @author Patrick Gebhardt
+ */
+@ExtendWith(FrameworkExtension.class)
+public class AlarmBeepServiceTest {
+
+	@Reference
+	private TestFramework framework;
+
+	private AlarmClockLogic acl;
```
Comment (sinpat) [inline plugins/alarmclock/src/test/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmBeepServiceTest.java]: Why is an AlarmClockLogic variable there? The alarmBeepService does not have an acl reference, so it is not needed. Also the test does not require it.
This is the case with a lot of the test classes.
You only have to declare one variable of 1. the class you are testing and 2. every class that is a reference of the tested class and of course other variables you need for the test
[Example](https://github.com/AmyAssist/Amy/blob/dev/plugins/email/src/test/java/de/unistuttgart/iaas/amyassist/amy/plugin/email/EmailLogicTest.java)

File: plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/Alarm.java [hunk 2]
```diff
@@ -79,28 +102,14 @@ public String toString() {
 	 * @return the corresponding alarm object
 	 */
 	public static Alarm reconstructObject(String input) {
```
Comment (Legion2) [inline plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/Alarm.java]: this method is not used/needed

File: plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/Alarm.java [hunk 3]
```diff
@@ -23,38 +23,61 @@
...
 /**
  * Class that defines timer attributes and behaviour
  * 
  * @author Patrick Singer, Patrick Gebhardt, Florian Bauer, Leon Kiefer
  */
-public class Alarm {
-
+@Entity
+@PersistenceUnit(unitName = "AlarmRegistry")
+public class Alarm implements RegistryEntity {
```
Comment (PassaChr) [inline plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/Alarm.java]: Should extend Entity and set the link object in the resource methods

File: plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmClockResource.java [hunk 4]
```diff
@@ -73,16 +70,8 @@
...
-		Timestamp[] timestamps = new Timestamp[alarms.size()];
-		for (int i = 0; i < alarms.size(); i++) {
-			if (alarms.get(i) != null) {
-				timestamps[i] = new Timestamp(alarms.get(i));
-				timestamps[i].setLink(createAlarmPath(alarms.get(i).getId()));
-			}
-		}
-		return timestamps;
+	public List<Alarm> getAllAlarms() {
+		return this.logic.getAllAlarms();
```
Comment (PassaChr) [inline plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmClockResource.java]: each alarm should have a link set that a client can use to access each alarm individually

File: plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmClockResource.java [hunk 5]
```diff
@@ -95,83 +84,96 @@
...
 		try {
 			alarm = this.logic.getAlarm(alarmnumber);
 		} catch (NoSuchElementException e) {
 			throw new WebApplicationException("there is no alarm" + alarmnumber, e, Status.NOT_FOUND);
 		}
 
-		Timestamp ts = new Timestamp(alarm);
-		ts.setLink(createAlarmPath(alarmnumber));
-		return ts;
+		return alarm;
```
Comment (PassaChr) [inline plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmClockResource.java]: s.o. for consistency

File: plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmClockResource.java [hunk 6]
```diff
@@ -95,83 +84,96 @@
...
-			return alarmTime;
+	public Alarm newAlarm(Alarm alarm) {
+		int day;
+		if (LocalDateTime.now().getDayOfMonth() == alarm.getAlarmTime().getDayOfMonth()) {
+			day = -1;
+		} else {
+			day = 1;
 		}
-		throw new WebApplicationException("The given time wasn't a valid time", Status.BAD_REQUEST);
+		return this.logic.setAlarm(day, alarm.getAlarmTime().getHour(), alarm.getAlarmTime().getMinute());
```
Comment (PassaChr) [inline plugins/alarmclock/src/main/java/de/unistuttgart/iaas/amyassist/amy/plugin/alarmclock/AlarmClockResource.java]: also set a link here
catch the exception thrown by the logic and rethrow the old exception

**Pull request merged**